internal/datastore/common/changes.go

```diff
@@ -4,99 +4,218 @@ import (
+			make(map[string]struct{}),
 		}
 		ch.records[k] = revisionChanges
 	}
 
-	tplKey := tuple.StringWithoutCaveat(tpl)
+	return revisionChanges
+}
 
-	switch op {
-	case core.RelationTupleUpdate_TOUCH:
-		// If there was a delete for the same tuple at the same revision, drop it
-		delete(revisionChanges.tupleDeletes, tplKey)
+// AddDeletedNamespace adds a change indicating that the namespace with the name was deleted.
+func (ch *Changes[R, K]) AddDeletedNamespace(
+	_ context.Context,
+	rev R,
+	namespaceName string,
+) {
+	if ch.content&datastore.WatchSchema != datastore.WatchSchema {
+		return
+	}
 
-		revisionChanges.tupleTouches[tplKey] = tpl
+	record := ch.recordForRevision(rev)
+	delete(record.definitionsChanged, nsPrefix+namespaceName)
 
-	case core.RelationTupleUpdate_DELETE:
-		_, alreadyTouched := revisionChanges.tupleTouches[tplKey]
-		if !alreadyTouched {
-			revisionChanges.tupleDeletes[tplKey] = tpl
-		}
+	record.namespacesDeleted[namespaceName] = struct{}{}
+}
+
+// AddDeletedCaveat adds a change indicating that the caveat with the name was deleted.
+func (ch *Changes[R, K]) AddDeletedCaveat(
+	_ context.Context,
+	rev R,
+	caveatName string,
+) {
+	if ch.content&datastore.WatchSchema != datastore.WatchSchema {
+		return
+	}
+
+	record := ch.recordForRevision(rev)
+	delete(record.definitionsChanged, caveatPrefix+caveatName)
+
+	record.caveatsDeleted[caveatName] = struct{}{}
+}
+
+// AddChangedDefinition adds a change indicating that the schema definition (namespace or caveat)
+// was changed to the definition given.
+func (ch Changes[R, K]) AddChangedDefinition(
+	ctx context.Context,
+	rev R,
+	def datastore.SchemaDefinition,
+) {
+	if ch.content&datastore.WatchSchema != datastore.WatchSchema {
+		return
+	}
+
+	record := ch.recordForRevision(rev)
+
+	switch t := def.(type) {
+	case *core.NamespaceDefinition:
+		delete(record.namespacesDeleted, t.Name)
+		record.definitionsChanged[nsPrefix+t.Name] = t
+
+	case *core.CaveatDefinition:
+		delete(record.caveatsDeleted, t.Name)
+		record.definitionsChanged[caveatPrefix+t.Name] = t
 	default:
-		log.Ctx(ctx).Fatal().Stringer("operation", op).Msg("unknown change operation")
+		log.Ctx(ctx).Fatal().Msg("unknown schema definition kind")
 	}
 }
 
 // AsRevisionChanges returns the list of changes processed so far as a datastore watch
 // compatible, ordered, changelist.
-func (ch Changes[R, K]) AsRevisionChanges(lessThanFunc func(lhs, rhs K) bool) []datastore.RevisionChanges {
+func (ch *Changes[R, K]) AsRevisionChanges(lessThanFunc func(lhs, rhs K) bool) []datastore.RevisionChanges {
+	return ch.revisionChanges(lessThanFunc, *new(R), false)
+}
+
+// FilteredRevisionChanges returns a list of changes processed up to the bound revision. Returns changes
+// are removed from the tracker.
+func (ch *Changes[R, K]) FilteredRevisionChanges(lessThanFunc func(lhs, rhs K) bool, boundRev R) []datastore.RevisionChanges {
+	changes := ch.revisionChanges(lessThanFunc, boundRev, true)
```

Comment:
Meh; most of the time the number of changes per revision should be no more than 1000 (since that's the limit), so I don't know if it is worth doing here.